[iOS] Entering fullscreen from iframe without explicit viewport results in badly cropped video

jernoble opened the pull request:
#### 66ba699e69102c7efbd92303666a45a23ab74bb2
<pre>
[iOS] Entering fullscreen from iframe without explicit viewport results in badly cropped video
<a href="https://bugs.webkit.org/show_bug.cgi?id=270909">https://bugs.webkit.org/show_bug.cgi?id=270909</a>
<a href="https://rdar.apple.com/123725878">rdar://123725878</a>

Reviewed by Abrar Rahman Protyasha.

In 270199@main, support was added for using the viewport settings from the outermost fullscreen
document, rather than the top document. This allowed an &lt;iframe&gt; whose contents specified a
viewport-fit=cover behavior to work within a hosting document without that viewport setting.
However, for an &lt;iframe&gt; whose contents did not specify a viewport at all, it would receive
the default desktop viewport, which is 980px wide. This, combined with the iOS fullscreen
behavior of blocking zoom, meant that the fullscreen viewport would be much, much larger than
the physical viewport.

When entering or exiting fullscreen mode, notify the WebPage, and reset the default viewport
such that in fullscreen the web page will receive the &quot;native&quot; viewport rather than the &quot;desktop&quot;
one by default. This requires storing the results of didReceiveMobileDocType(), so that it
can be re-used after page load.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::willEnterFullScreen):
(WebKit::WebFullScreenManager::didExitFullScreen):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::isInFullscreenChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::didReceiveMobileDocType):
(WebKit::WebPage::resetViewportDefaultConfiguration):

Canonical link: <a href="https://commits.webkit.org/276138@main">https://commits.webkit.org/276138@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f7a0c50e654017b1ba5a4f203c490b931ab5caa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22862 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46249 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46457 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39908 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26746 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20268 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36168 "8 flakes 18 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19900 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37736 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17162 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17424 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38813 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1867 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39990 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39115 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48015 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18837 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42976 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20233 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38007 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41679 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9754 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20433 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19857 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -522,6 +522,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 
 #if ENABLE(FULLSCREEN_API)
     WebFullScreenManager* fullScreenManager();
+    void isInFullscreenChanged(bool);
```
nit: Maybe we make this argument an `enum class IsInFullscreenMode : bool { No, Yes }` to improve call site legibility.

nt1m commented on LayoutTests/fast/viewport/ios/resources/no-viewport.html:
```diff
@@ -0,0 +1,12 @@
+<script>
+window.addEventListener('load', event => {
+    document.body.onclick = function() {
+        document.body.webkitRequestFullScreen();
```
Prefix is probably not needed

jernoble committed:
[iOS] Entering fullscreen from iframe without explicit viewport results in badly cropped video
https://bugs.webkit.org/show_bug.cgi?id=270909
rdar://123725878

Reviewed by Abrar Rahman Protyasha.

In 270199@main, support was added for using the viewport settings from the outermost fullscreen
document, rather than the top document. This allowed an <iframe> whose contents specified a
viewport-fit=cover behavior to work within a hosting document without that viewport setting.
However, for an <iframe> whose contents did not specify a viewport at all, it would receive
the default desktop viewport, which is 980px wide. This, combined with the iOS fullscreen
behavior of blocking zoom, meant that the fullscreen viewport would be much, much larger than
the physical viewport.

When entering or exiting fullscreen mode, notify the WebPage, and reset the default viewport
such that in fullscreen the web page will receive the "native" viewport rather than the "desktop"
one by default. This requires storing the results of didReceiveMobileDocType(), so that it
can be re-used after page load.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::willEnterFullScreen):
(WebKit::WebFullScreenManager::didExitFullScreen):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::isInFullscreenChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::didReceiveMobileDocType):
(WebKit::WebPage::resetViewportDefaultConfiguration):

Canonical link: https://commits.webkit.org/276138@main
